README.md
# NewsMapTracker

This project is a news dashboard that displays up-to-date news in real time on a world map based 
on location of occurrence. Every hour, a new set of markers are displayed on the map, each pinpointing the location of a recent news event. Upon clicking a marker,
you will see the headline for that event as well as a link to the news article.

<img width="1382" alt="Screen Shot 2023-09-30 at 11 54 39 PM" src="https://github.com/gmantri2/NewsMapTracker/assets/84867535/b16f70ce-06a0-4e4f-8889-a441fb55f20f">


news.txt
[{"title":"Boats Missing in Spain: About 300 Migrants Traveling in Three Boats From Senegal Disappear at Sea Near Spanish Canary Islands","link":"https://www.latestly.com/socially/world/boats-missing-in-spain-about-300-migrants-traveling-in-three-boats-from-senegal-disappear-at-sea-near-spanish-canary-islands-5254228.html","keywords":null,"creator":null,"video_url":null,"description":"The news was confirmed by the migrant aid group Walking Borders. Helena Maleno of Walking Borders said that two boats, one carrying about 65 people and the other with between 50 and 60 on board, have been missing for 15 days since they left Senegal to try to reach Spain.","content":"At least 300 people are said to be missing after their boats disappeared while trying to reach Spain's Canary Islands. As per a report in Reuters, the people were travelling on three migrant boats from Senegal to Spain's Canary Islands before they and their boats disappeared. The news was confirmed by the migrant aid group Walking Borders. Helena Maleno of Walking Borders said that two boats, one carrying about 65 people and the other with between 50 and 60 on board, have been missing for 15 days since they left Senegal to try to reach Spain. A third boat left for Senegal on June 27 with about 200 people aboard. EU Commissioners Arrive in Madrid to Discuss Plans for Spain's EU Presidency. Migrants Missing at Sea Near Spanish Canary Islands BREAKING: Over 300 migrants missing at sea near Spanish Canary Islands — The Spectator Index (@spectatorindex) July 9, 2023 Over 300 Migrants Missing at Sea More than 300 migrants missing after their boats disappeared while trying to reach Spain's Canary Islands - Reuters — BNO News (@BNONews) July 9, 2023 (SocialLY brings you all the latest breaking news, viral trends and information from social media world, including Twitter, Instagram and Youtube. The above post is embeded directly from the user's social media account and LatestLY Staff may not have modified or edited the content body. The views and facts appearing in the social media post do not reflect the opinions of LatestLY, also LatestLY does not assume any responsibility or liability for the same.)","pubDate":"2023-07-10 01:46:26","image_url":null,"source_id":"latestly","category":["world"],"country":["india"],"language":"english"},{"title":"‘Life threatening’ flooding overwhelms New York roadways, killing 1 person","link":"https://www.winnipegfreepress.com/arts-and-life/life/greenpage/2023/07/09/life-threatening-flooding-overwhelms-new-york-roadways-killing-1-person","keywords":["Environment","World"],"creator":["Feedloaderapi"],"video_url":null,"description":"HUDSON VALLEY, N.Y. (AP) — Heavy rain sparked extreme flooding in New York’s Hudson Valley that killed at least one person, swamped roadways and forced road closures on Sunday night. […]","content":"HUDSON VALLEY, N.Y. (AP) — Heavy rain sparked extreme flooding in New York's Hudson Valley that killed at least one person, swamped roadways and forced road closures on Sunday night. Read this article for free: Already have an account? To continue reading, please subscribe: + GST for 9 weeks of unlimited access* HUDSON VALLEY, N.Y. (AP) — Heavy rain sparked extreme flooding in New York’s Hudson Valley that killed at least one person, swamped roadways and forced road closures on Sunday night. The National Weather Service issued flash flood warnings across parts of southeastern New York, describing it as “life threatening.” One person died due to the flooding, Steven M. Neuhaus, the Orange County executive, told The New York Times. State Route 9W was flooded, and the Palisades Interstate Parkway became so drenched that parts of it was closed, the New York State Police said in a statement. The police asked the public to avoid the parkway. Cedar Pond Brook in Stony Point was flowing over the road and into private properties, according to WABC. Rockland County Executive Ed Day instructed residents to “remain indoors in a safe location” until the heavy rainfall ended. Advertisement","pubDate":"2023-07-10 01:34:21","image_url":null,"source_id":"winnipegfreepress","category":["world"],"country":["canada"],"language":"english"},{"title":"United States: Supreme Court refuses to hear challenge to Mississippi law against Black voting rights","link":"https://www.greenleft.org.au/content/united-states-supreme-court-refuses-hear-challenge-mississippi-law-against-black-voting","keywords":null,"creator":["pricesusan9@gmail.com"],"video_url":null,"description":"The United States Supreme Court upheld, on June 30, a century-old Mississippi law used during the “Jim Crow” segregationist era to deny Black people in the state the right to vote, reports Malik Miah.","content":"So much for the “colour blind” society proclaimed by the Supreme Court in its ruling June 30 to outlaw affirmative action policies at public universities. The same day, but under the radar of the mainstream news media and public, the unelected body with lifetime terms made its full objective crystal clear. The six far-right majority Justices of the Court upheld a century-old Mississippi amendment to its state constitution that was used during the “Jim Crow” segregationist era to deny Black people in the state the right to vote and fully participate as equals. The president of the 1890 constitutional convention that adopted the amendment made it clear, saying: “We came here to exclude the Negro.” By a six-to-three vote, the Court rejected hearing a challenge to the amendment. Four Justices are required to vote in favour for a case to be heard The state argued that the provision is no longer tainted by the racist intentions of its original authors. Why then leave it in the constitution? The court's decision prompted a sharp dissenting opinion from Justice, Ketanji Brown Jackson — the first Black woman appointed to the Court. She was joined by liberal Justice Sonia Sotomayor, the first Latina on the court. Jackson exposed the Court majo

public/app.js
// ----------map stuff

// const PORT = process.env.PORT || 3030;

const container = document.getElementById('popup');
const content = document.getElementById('popup-content');
const closer = document.getElementById('popup-closer');

const key = 'FAvR4BNiT6kBQVkKBpE4';

const overlay = new ol.Overlay({
    element: container,
    autoPan: {
      animation: {
        duration: 250,
      },
    },
  });

closer.onclick = function () {
    overlay.setPosition(undefined);
    closer.blur();
    return false;
};

const attribution = new ol.control.Attribution({
    collapsible: false,
});

const source = new ol.source.TileJSON({
    url: `https://api.maptiler.com/maps/streets-v2/tiles.json?key=${key}`, // source URL
    tileSize: 512,
    crossOrigin: 'anonymous'
});

const map = new ol.Map({
layers: [
    new ol.layer.Tile({
        source: source
    })
],
controls: ol.control.defaults.defaults({attribution: false}).extend([attribution]),
    target: 'map',
    view: new ol.View({
        constrainResolution: false,
        center: ol.proj.fromLonLat([10, 25]), // starting position [lng, lat]
    zoom: 1 // starting zoom
    }),
overlays: [overlay],
});

var commonWords = new Set([
    "Police",
    "Best",
    "Sale",
    "Most",
    "Man",
    "Date",
    "Chelsea",
    "Opportunity",
    "Bo",
    "Mobile",
    "Liberty",
    "Mission",
    "Summit",
    "Asia",
    "Washington"
  ]);

function filter_words(cities) {
    let new_list = cities.filter(function(item)
    {
         return !(commonWords.has(item));
    });
    return new_list;
}

function helper(text) {
    for (let i = 1; i < text.length; i++) {
        if (text[i-1] != ' ') {
            text[i] = text[i].toLowerCase()
        }
    }
    return text.join('');
}

async function import_geotext() {
    const pyodideRuntime = await loadPyodide();
    await pyodideRuntime.loadPackage("micropip");
    const micropip = pyodideRuntime.pyimport("micropip");
    await micropip.install("geotext");
    pyodideRuntime.runPython(`
      import js
      from geotext import GeoText
      def get_locations(sample_text):
        places = GeoText(sample_text)
        return places.cities
      js.get_locations = get_locations
    `);
  }

async function get_locations_list(news_text) {
    news_text = news_text.replace(/['"]+/g, '');
    var output = (Array.from(get_locations(news_text)))
    return output
  }

function create_marker(coord) {
    const layer = new ol.layer.Vector({
        source: new ol.source.Vector({
            features: [
            new ol.Feature({
                geometry: new ol.geom.Point(ol.proj.fromLonLat(coord)),
            })
            ]
        }),
        style: new ol.style.Style({
            image: new ol.style.Icon({
            anchor: [0.5, 1],
            crossOrigin: 'anonymous',
            src: 'https://docs.maptiler.com/openlayers/default-marker/marker-icon.png',
            })
        })
    });
    return layer;
}

async function getInfo() {
    const res = await fetch(`https://news-map-tracker.onrender.com/info`, {
        method: 'GET'
    })
    // const res = await fetch('https://b672-2600-4040-2086-c700-c9e6-4c46-abf9-5a98.ngrok-free.app/info', {
    //     method: 'GET'
    // })
    const news = await res.json()
    console.log(news)

    var layers = [];
    var urls = [];
    var titles = [];
    await import_geotext();
    for (let i = 0; i < news.length; i++) {
        var text = news[i].title
        var url = news[i].link

        var description = news[i].description
        if (description) {
            description_processed = helper(description.split(''))
            var result = await get_locations_list(description_processed);
        } else {
            var result = await get_locations_list(text);
        }

        //var result = await get_locations_list(text);

        result = filter_words(result);
        if (result.length == 1 || (new Set(result)).size == 1) { //can change result criteria
            console.log(text)
            console.log(result)
            var loc = result[0]
            console.log(loc, i)
            var query2 = "&limit=5"
            var response2 = await fetch(`https://api.maptiler.com/geocoding/` + loc + `.json?key=${key}` + query2);
            var result2 = await response2.json();
            var coord = result2.features[0].center
            console.log(coord)

            const marker = create_marker(coord);
            map.addLayer(marker);

            console.log(url)
            layers.push(marker);
            urls.push(url);
            titles.push(text);
        }
    }

    map.on('singleclick', function (evt) {
        var rets = map.forEachFeatureAtPixel(evt.pixel, function(feature, clicked_layer) {
            for (var i = 0; i < layers.length; i++) {
              if (clicked_layer == layers[i]) {
                return [feature, i]
              }
            }
        });
        if (rets) {
            const url = urls[rets[1]];
            const title = titles[rets[1]]
            var url_ref = title.link(url);

            var position = 3;
            var url_ref_new = [url_ref.slice(0, position), "target=\"_blank\" ", url_ref.slice(position)].join('');

            content.innerHTML = url_ref_new;
            const coordinate = evt.coordinate;
            overlay.setPosition(coordinate);
        }
      });
}

getInfo();

index.js
const express = require("express");

const app = express()
const PORT = process.env.PORT || 3030;

app.use(express.static('public'))

app.listen(PORT, () => console.log(`Server running on port ${PORT}`))

async function update_news() {

    var key = 'pub_2591697c6a94d438b79875dbcdae7c1f58443';
    var categories = 'world'
    var response4 = await fetch(`https://newsdata.io/api/1/news?apikey=${key}&language=en&category=${categories}`);
    var result4 = await response4.json();
    var result4news = result4.results
    var nextPageCode = result4.nextPage

    for (var i = 0; i < 10; i++) {
        if (nextPageCode) {
            var response5 = await fetch(`https://newsdata.io/api/1/news?apikey=${key}&language=en&category=${categories}&page=${nextPageCode}`)
            var result5 = await response5.json();
            var result5news = result5.results
            nextPageCode = result5.nextPage
            result4news = result4news.concat(result5news)
        }
    }

    //write to file
    var fs = require('fs');
    fs.writeFile("latest_news.txt", JSON.stringify(result4news), function(err) {
        if (err) {
            console.log(err);
        }
    });
}

// create GET route on on express server API 
app.get("/info", (req, res) => {
    var fs = require('fs');
    var result4news = JSON.parse(fs.readFileSync("latest_news.txt"))
    res.send(result4news)
})

var CronJob = require('cron').CronJob;
var job = new CronJob(
    '30 7-23 * * *',
    function() {
        console.log("updating news...");
        update_news()
    },
    null,
    true,
    'America/New_York'
);

const request = require('request');
var job = new CronJob(
    '*/10 * * * *',
    function() {
        request('https://news-map-tracker.onrender.com/', (error, response, body)=>{
            console.log("pinging...");
        })
    },
    null,
    true,
    'America/New_York'
);